### PR TITLE
abci: Synchronize FinalizeBlock with the updated specification

### PR DESCRIPTION
**Pull request by williambanfield**
This change set implements the most recent version of `FinalizeBlock`. 

# What does this change actually contain?

* This change set is rather large but fear not! The majority of the files touched and changes are renaming `ResponseDeliverTx` to `ExecTxResult`. This should be a pretty inoffensive change since they're effectively the same type but with a different name.
* The `execBlockOnProxyApp` was totally removed since it served as just a wrapper around the logic that is now mostly encapsulated within `FinalizeBlock`
* The `updateState` helper function has been made a public method on `State`. It was being exposed as a shim through the testing infrastructure, so this seemed innocuous.
* Tests already existed to ensure that the application received the `ByzantineValidators` and the `ValidatorUpdates`, but one was fixed up to ensure that `LastCommitInfo` was being sent across.
* Tests were removed from the `psql` indexer that seemed to search for an event in the indexer that was not being created.

# Questions for reviewers
* We store this [ABCIResponses](https://github.com/tendermint/tendermint/blob/5721a13ab1f4479f9807f449f0bf5c536b9a05f2/proto/tendermint/state/types.pb.go#L37) type in the data base as the block results. This type has changed since v0.35 to contain the `FinalizeBlock` response. I'm wondering if we need to do any shimming to keep the old data retrieveable?
* Similarly, this change is exposed via the RPC through [ResultBlockResults](https://github.com/tendermint/tendermint/blob/5721a13ab1f4479f9807f449f0bf5c536b9a05f2/rpc/coretypes/responses.go#L69) changing. Should we somehow shim or notify for this change? 

closes: #7658

### CODE REVIEW

File: internal/consensus/mempool_test.go [hunk 1]
```diff
@@ -118,15 +118,15 @@ func TestMempoolProgressInHigherRound(t *testing.T) {
 	round = 0
 
 	ensureNewRound(t, newRoundCh, height, round) // first round at next height
-	deliverTxsRange(ctx, t, cs, 0, 1)            // we deliver txs, but dont set a proposal so we get the next round
+	checkTxsRange(ctx, t, cs, 0, 1)              // we deliver txs, but dont set a proposal so we get the next round
```
Comment (creachadair) [inline internal/consensus/mempool_test.go:121]:
```suggestion
	checkTxsRange(ctx, t, cs, 0, 1)              // we deliver txs, but don't set a proposal so we get the next round
```

File: internal/state/execution_test.go [hunk 2]
```diff
@@ -68,8 +67,10 @@ func TestApplyBlock(t *testing.T) {
 	assert.EqualValues(t, 1, state.Version.Consensus.App, "App version wasn't updated")
 }
 
-// TestBeginBlockValidators ensures we send absent validators list.
-func TestBeginBlockValidators(t *testing.T) {
+// TestFinalizeBlockLastCommitInfo ensures we correctly send the LastCommitInfo to the
+// applications. The test ensures that the LastCommitInfo properly reflects the
```
Comment (creachadair) [inline internal/state/execution_test.go:71]:
```suggestion
// application. The test ensures that the LastCommitInfo properly reflects
```

File: internal/state/execution_test.go [hunk 3]
```diff
@@ -80,67 +81,59 @@ func TestBeginBlockValidators(t *testing.T) {
...
-		absentSig = types.NewCommitSigAbsent()
-	)
+	absentSig := types.NewCommitSigAbsent()
 
 	testCases := []struct {
-		desc                     string
-		lastCommitSigs           []types.CommitSig
-		expectedAbsentValidators []int
+		name             string
+		absentCommitSigs map[int]struct{}
```
Comment (creachadair) [inline internal/state/execution_test.go:90]: Not crucial, but it would be fine to use a map to bool here simply for legibility. Using empty structs is mainly relevant for large sets where the value size may matter. YMMV, though.

File: internal/state/execution_test.go [hunk 4]
```diff
@@ -80,67 +81,59 @@ func TestBeginBlockValidators(t *testing.T) {
...
 
-				assert.False(t, v.SignedLastBlock)
-				ctr++
-			} else {
-				assert.True(t, v.SignedLastBlock)
+			// block for height 2
+			block, err := sf.MakeBlock(state, 2, lastCommit)
+			require.NoError(t, err)
+
+			require.NoError(t, err)
```
Comment (creachadair) [inline internal/state/execution_test.go:116]: Duplicate?

File: internal/state/execution_test.go [hunk 5]
```diff
@@ -80,67 +81,59 @@ func TestBeginBlockValidators(t *testing.T) {
...
+			bps, err := block.MakePartSet(testPartSize)
+			require.NoError(t, err)
+			blockID := types.BlockID{Hash: block.Hash(), PartSetHeader: bps.Header()}
+			_, err = blockExec.ApplyBlock(ctx, state, blockID, block)
+			require.NoError(t, err)
+
+			// -> app receives a list of validators with a bool indicating if they signed
+			for i, v := range app.CommitVotes {
+				if _, ok := tc.absentCommitSigs[i]; ok {
```
Comment (creachadair) [inline internal/state/execution_test.go:125]: Alternatively: 

```go
_, absent := tc.absentCommitSigs[i]
assert.Equal(t, !absent, v.SignedLastBlock)
```

File: internal/state/indexer/block/kv/kv.go [hunk 6]
```diff
@@ -65,19 +64,19 @@ func (idx *BlockerIndexer) Index(bh types.EventDataNewBlockHeader) error {
...
 	return batch.WriteSync()
 }
 
-// Search performs a query for block heights that match a given BeginBlock
-// and Endblock event search criteria. The given query can match against zero,
-// one or more block heights. In the case of height queries, i.e. block.height=H,
-// if the height is indexed, that height alone will be returned. An error and
-// nil slice is returned. Otherwise, a non-nil slice and nil error is returned.
+// Search performs a query for block heights that match a given FinalizeBlock
+// The given query can match against zero, one or more block heights. In the case
```
Comment (creachadair) [inline internal/state/indexer/block/kv/kv.go:76]:
```suggestion
// The given query can match against zero or more block heights. In the case
```

**Review by creachadair — APPROVED** [review]
You were right, this wasn't so scary. 🙂

As regarding the ABCI responses, I think we are talking about getting rid of them from the store entirely. They take up a lot of space and are not exposed via any API of the node. 

On that basis I don't think we need to worry about compatibility even if we do keep the data.

**Comment by creachadair** [comment]
Regarding the RPC: Does the JSON structure of the block result change much? My sense is that this shouldn't matter too much, but we might want to check if relayers rely on it. 

That reminds me, when we do update the protos we should also make sure the JSON marshal wrappers get updated too. I have been meaning to make sure that gets enforced in CI but haven't done so yet.

File: internal/state/execution_test.go [hunk 7]
```diff
@@ -80,67 +81,59 @@ func TestBeginBlockValidators(t *testing.T) {
...
-		lastCommitSigs           []types.CommitSig
-		expectedAbsentValidators []int
+		name             string
+		absentCommitSigs map[int]bool
 	}{
-		{"none absent", []types.CommitSig{commitSig0, commitSig1}, []int{}},
-		{"one absent", []types.CommitSig{commitSig0, absentSig}, []int{1}},
-		{"multiple absent", []types.CommitSig{absentSig, absentSig}, []int{0, 1}},
+		{"none absent", map[int]bool{}},
+		{"one absent", map[int]bool{1: true}},
```
Comment (sergio-mena) [inline internal/state/execution_test.go:93]: Minor thing. Since we always have _true_ as the value, we could turn the map into an array of index here, couldn't we?
Besides, the field's name is already saying those are "absent"

File: internal/consensus/state_test.go [hunk 8]
```diff
@@ -1965,6 +1965,79 @@ func TestProcessProposalAccept(t *testing.T) {
...
+			startTestRound(ctx, cs1, cs1.Height, round)
+			ensureNewRound(t, newRoundCh, height, round)
+			ensureNewProposal(t, proposalCh, height, round)
+			rs := cs1.GetRoundState()
+
+			blockID := types.BlockID{}
+			nextRound := round + 1
+			nextHeight := height
+			if !testCase.voteNil {
+				polRound = 0
```
Comment (thanethomson) [inline internal/consensus/state_test.go:2017]:
```suggestion
```

**Comment by williambanfield** [comment]
> Regarding the RPC: Does the JSON structure of the block result change much? My sense is that this shouldn't matter too much, but we might want to check if relayers rely on it.

The only change to the structure is the `Events` field is now called `TxEvents`. The marshalled JSON results an object with field `tx_events` instead of `events`.

The relayer [does appear to make use of this type])https://github.com/cosmos/relayer/blob/ab6bc21cb1dccd8c49fabf5ecd710c1fa2a5df41/relayer/provider/provider.go#L30). It effectively implements its own version. 

In that case, our options are to 1. create a new wrapper type that marshals to the old field so that client remain compatible, 2. update the relayer to recognize the new field name, 3. use the old events field name. 
I'm partial to 3 for simplicity. The field name is also scoped to the within an object with `Tx` in the name, so I feel as though not much information is really gained by having the extra `tx_` prefix here, but I'm open to being convinced otherwise. 

> That reminds me, when we do update the protos we should also make sure the JSON marshal wrappers get updated too. I have been meaning to make sure that gets enforced in CI but haven't done so yet.

Anything i can do to help out with this?

**Comment by creachadair** [comment]
> The only change to the structure is the `Events` field is now called `TxEvents`. The marshalled JSON results an object with field `tx_events` instead of `events`.

Before I had read the next paragraph, I was going to suggest just keeping the original JSON field tag. So yeah, let's plan on that. This might be a little tedious because of the way protobuf names map to JSON, but we can shim it in the RPC if we have to.

> > That reminds me, when we do update the protos we should also make sure the JSON marshal wrappers get updated too. I have been meaning to make sure that gets enforced in CI but haven't done so yet.
> 
> Anything I can do to help out with this?

Right now, as far as I can tell, it's entirely manual. In `abci/types/result.go` we have some hand-written shims for JSON encoding on response types, which we might need to update. (Protobuf annoyingly does not generate these, and the standard encoding/json package does not correctly handle them out of the box). I've got a tool that can automate this, I just need to kick some dust off it and wire it up.

**Comment by williambanfield** [comment]
TODO:
- [] Update `addTxCmd` to `finalizeBlockCmd` so that users can create a 'finalize block'
- [] handle `tx_event` encoding change to use the json encoding fields.

File: internal/consensus/state_test.go [hunk 9]
```diff
@@ -1965,6 +1965,79 @@ func TestProcessProposalAccept(t *testing.T) {
...
+			ensurePrevoteMatch(t, voteCh, height, round, rs.ProposalBlock.Hash())
+
+			signAddVotes(ctx, t, cs1, tmproto.PrecommitType, config.ChainID(), blockID, vss[1:]...)
+			ensurePrecommit(t, voteCh, height, round)
+
+			ensureNewRound(t, newRoundCh, nextHeight, nextRound)
+			m.AssertExpectations(t)
+			if testCase.voteNil {
+				m.AssertNotCalled(t, "FinalizeBlock", mock.Anything)
+			}
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go]: Thanks

File: internal/state/execution.go [hunk 10]
```diff
@@ -207,38 +207,45 @@ func (blockExec *BlockExecutor) ValidateBlock(ctx context.Context, state State,
...
+			ByzantineValidators: block.Evidence.ToABCI(),
+		},
 	)
 	endTime := time.Now().UnixNano()
 	blockExec.metrics.BlockProcessingTime.Observe(float64(endTime-startTime) / 1000000)
 	if err != nil {
 		return state, ErrProxyAppConn(err)
 	}
 
+	abciResponses := &tmstate.ABCIResponses{
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/state/execution.go:233]: Yeah, separate PR is best

File: internal/state/execution_test.go (hunk 7 above)
Comment (sergio-mena) [reply internal/state/execution_test.go:93]: So, I was thinking of something like:
```go
	{"none absent", []int{}},
	{"one absent", []int{1}},
	{"multiple absent", []int{1, 3}},
```
but then I read again through the code and line 125 below makes it sub-optimal w.r.t. the map-based solution (the array-based solution would need a nested *for* loop).

**Comment by williambanfield** [comment]
@sergio-mena, Before I go creating a type to wrap the `ExecTxResult` fields so as to keep the old `json` field name, I'd like to get your input: It appears we should keep the same field name of the `tx_events` field, which used to be called `events`. This field is used by the relayer, so it looks like we should keep the name the same so as to not break our users. 

I'm wondering if you feel strongly about keeping the internal type name as `TxEvents`? 

The field is within a `ExecTxResult` so I'm not sure we should be too worried about the events being ambiguous in some way if we just dropped the `Tx` from the field name. This would allow us to keep the old `json` encoding without having to do any shimming. If you feel strongly about the new name, I can definitely wire up a shim.

**Comment by sergio-mena** [comment]
> @sergio-mena, Before I go creating a type to wrap the `ExecTxResult` fields so as to keep the old `json` field name, I'd like to get your input: It appears we should keep the same field name of the `tx_events` field, which used to be called `events`. This field is used by the relayer, so it looks like we should keep the name the same so as to not break our users.
> 
> I'm wondering if you feel strongly about keeping the internal type name as `TxEvents`?
> 
> The field is within a `ExecTxResult` so I'm not sure we should be too worried about the events being ambiguous in some way if we just dropped the `Tx` from the field name. This would allow us to keep the old `json` encoding without having to do any shimming. If you feel strongly about the new name, I can definitely wire up a shim.

My view is that, if we risk breaking IBC due to renaming "events" to "tx_events", it's not worth it. So I think I'm aligned with you on this.

On a somewhat related point. I decided on the `ExecTxResult` name because `TxResult` was already existing (this was the name I wanted to use initially). I remember doing an extensive search of `TxResult` and the situation looked a bit strange to me: `TxResult` was defined as part of ABCI, but ABCI didn't use it anywhere. IIRC, it was only used somewhere in the rpc code.
In summary, I think it would be cool if we could get rid of current `TxResult` (e.g., by moving it away from ABCI protobuf definitions) and renaming `ExecTxResult` to `TxResult`.

**Comment by williambanfield** [comment]
> My view is that, if we risk breaking IBC due to renaming "events" to "tx_events", it's not worth it. So I think I'm aligned with you on this. Nevertheless, if you rename it, please adapt the corresponding field name in the spec.

Switched to the old name, made sure to update the spec as well.

File: internal/state/execution.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (williambanfield) [reply internal/state/execution.go:233]: Opened https://github.com/tendermint/tendermint/issues/8069 to track.
